hw3/3.2/main.c

```diff
@@ -0,0 +1,171 @@
+    if (blockSize <= 0) {
+        return;
+    }
+    if (blockSize < 10) {
+        insertionSort(array, leftBound, rightBound);
+        return;
+    }
+
+    //chooses pivot
+    int pivot = array[leftBound];
+    bool areAllEqual = true;
+    for (int i = leftBound; i < rightBound; i++) {
+        if (array[i] != array[i + 1]) {
+            pivot = array[i] > array[i + 1] ? array[i] : array[i + 1];
+            areAllEqual = false;
+            break;
+        }
+    }
+
+    if (areAllEqual) {
+        return;
+    }
+
+    int leftPartIndex = 0;
+    int rightPartIndex = 0;
+    partition(array, leftBound, rightBound, &leftPartIndex, &rightPartIndex, pivot);
+    quickSort(array, leftBound, rightPartIndex);
+    quickSort(array, leftPartIndex, rightBound);
+}
+
+bool binSearch(const int *array, const int size, const int key) {
+    // leftBound <= key, rightBound > key
+    int leftBound = 0;
+    int rightBound = size;
+    while (rightBound - leftBound > 1) {
+        int middle = (leftBound + rightBound) / 2;
+        if (array[middle] <= key) {
+            leftBound = middle;
+        } else {
+            rightBound = middle;
+        }
+    }
+    return array[leftBound] == key;
+}
+
+bool doesExist(const int *array, const int size, const int key) {
+    for (int i = 0; i < size; i++) {
+        if (array[i] == key) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool testBinSearch(const int *array, const int size, const int key) {
+    return doesExist(array, size, key) == binSearch(array, size, key);
+}
+
+void generateRandomArray(int *array, const int size) {
+    for (int i = 0; i < size; i++) {
+        array[i] = rand() % TEST_ARRAY_SIZE;
+    }
+}
+
+bool isPassed() {
+    int testArray[TEST_ARRAY_SIZE] = {0};
+    int testKeys[TEST_ARRAY_SIZE] = {0};
+    generateRandomArray(testArray, TEST_ARRAY_SIZE);
+    generateRandomArray(testKeys, TEST_ARRAY_SIZE);
+
+    quickSort(testArray, 0, TEST_ARRAY_SIZE - 1);
+    bool result = true;
+    for (int i = 0; i < TEST_ARRAY_SIZE; i++) {
+        result &= testBinSearch(testArray, TEST_ARRAY_SIZE, testKeys[i]);
+    }
+    return result;
+}
+
+
+int main() {
+    setlocale(LC_ALL, "RU-ru");
+    srand(239);
+
+    if (!isPassed()) {
+        printf("тесты не пройдены\n");
+        return 0;
+    }
+
+    printf("введите размер массива и кол-во случайных элементов (n и k): \n");
+    int n = 0, k = 0;
+    scanf("%d%d", &n, &k);
+    while (n < 0 || k < 0) {
+        printf("n и k должны быть неотрицательными\n");
+        printf("введите размер массива и кол-во случайных элементов (n и k): \n");
+        scanf("%d%d", &n, &k);
+    }
+    int *randomArray = calloc(n, sizeof(int));
+    int *randomNumbers = calloc(k, sizeof(int));
+
+    for (int i = 0; i < n; i++) {
+        randomArray[i] = rand() % n;
+    }
+    for (int i = 0; i < k; i++) {
+        randomNumbers[i] = rand() % n;
+    }
+
+    quickSort(randomArray, 0, n - 1);
+
+    printf("массив:\n");
```

Comment:
```suggestion
    printf("Отсортированный массив:\n");
```
Не надо пользователя в заблуждение вводить. Без дополнительного уточнения кажется, что массив случайно сгенерировался отсортированным.